tests/ghost_test.go

```diff
@@ -0,0 +1,346 @@
+	migrationContext.SetMaxLagMillisecondsThrottleThreshold(*maxLagMillis)
+	migrationContext.SetThrottleQuery(*throttleQuery)
+	migrationContext.SetThrottleHTTP(*throttleHTTP)
+	migrationContext.SetIgnoreHTTPErrors(*ignoreHTTPErrors)
+	migrationContext.SetDefaultNumRetries(*defaultRetries)
+	migrationContext.ApplyCredentials()
+	if err := migrationContext.SetupTLS(); err != nil {
+		return err
+	}
+	if err := migrationContext.SetCutOverLockTimeoutSeconds(*cutOverLockTimeoutSeconds); err != nil {
+		return err
+	}
+	if err := migrationContext.SetExponentialBackoffMaxInterval(*exponentialBackoffMaxInterval); err != nil {
+		return err
+	}
+
+	migrator := logic.NewMigrator(migrationContext)
+	err := migrator.Migrate()
+	if err != nil {
+		_ = migrator.ExecOnFailureHook()
+		return err
+	}
+	fmt.Fprintf(os.Stdout, "# Done\n")
+	return nil
+}
+
+func TestGhostSimpleNoop(t *testing.T) {
+	basedir := t.TempDir()
+	datadir := filepath.Join(basedir, "data")
+	if err := os.Mkdir(datadir, 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	mysql, err := mysql.Install(basedir, datadir, "root")
+	if err != nil {
+		t.Fatalf("failed to start MySQL: %v", err)
+	}
+	if err := mysql.Start(13306, os.Stdout, os.Stderr, 60); err != nil {
+		t.Fatalf("failed to start MySQL: %v", err)
+	}
+
+	db, err := sql.Open("mysql", fmt.Sprintf("root@tcp(localhost:%d)/mysql", mysql.Port()))
+	if err != nil {
+		t.Fatalf("failed to open MySQL: %v", err)
+	}
+	defer db.Close()
+
+	_, err = db.Exec("CREATE DATABASE gh_ost_test_db")
+	if err != nil {
+		t.Fatalf("failed to CREATE DATABASE gh_ost_test_db")
+	}
+
+	_, err = db.Exec("USE gh_ost_test_db")
+	if err != nil {
+		t.Fatalf("failed to USE gh_ost_test_db")
+	}
+
+	_, err = db.Exec("CREATE TABLE tbl (id int primary key, data int)")
+	if err != nil {
+		t.Fatalf("failed to CREATE TABLE tbl (id int primary key, data int)")
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		t.Fatalf("failed to start a transaction: %v", err)
+	}
+	for i := 1; i <= 100000; i++ {
+		_, err := tx.Exec(fmt.Sprintf("INSERT INTO tbl values (%v, %v)", i, i))
+		if err != nil {
+			t.Fatalf("failed to insert: %v", err)
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		t.Fatalf("failed to commit: %v", err)
+	}
+
+	err = ghostRun([]string{"gh-ost", "--host=127.0.0.1", "--port=13306", "--allow-on-master", "--user=root", "--database=gh_ost_test_db", "--table=tbl", "--alter=add name varchar(64)"})
+
+	if err != nil {
+		t.Fatalf("failed to run gh-ost: %v", err)
+	}
+
+	rows, err := db.Query(`SELECT * FROM tbl`)
+	if err != nil {
+		t.Fatalf("failed to SELECT * FROM tbl: %v", err)
+	}
+
+	var (
+		id   int
+		data int
+	)
+
+	for rows.Next() {
+		if err := rows.Scan(&id, &data); err != nil {
+			t.Fatalf("failed to scan: %v", err)
+		}
+		if id != data {
+			t.Errorf("data mismatch, expect id: %v, data: %v, get id: %v, data: %v", id, id, id, data)
+		}
+	}
+	rows.Close()
+
+	if err := mysql.Stop(os.Stdout, os.Stderr); err != nil {
```

Comment:
May be this should be also called via defer, otherwise mysql won't be closed if abort prematurely?